### DISCUSSION TITLE
Q: 关于 hook || Q: About hooks

### DISCUSSION THREAD

**Discussion by felixesintot** [walinejs/waline#2096] [Q&A] [answered]
### Describe the bug | 问题描述

想实现两个简单的功能：要求评论的昵称和邮箱不为空以及注册用户的昵称不能与他人一样

但不是很懂 hook 怎么写，有没有大佬教教我

### Website URL | 问题网站

https://waline.felixesintot.top

### Where your waline deploy? | 服务部署在哪里？

Vercel (Default)

### Where your comment data store? | 数据存储在哪里？

LeanCloud(https://leancloud.app)

---
### Describe the bug | Problem description

I want to implement two simple functions: the nickname and email address of the comment must not be empty, and the nickname of the registered user cannot be the same as others.

But I don’t know how to write hook. Can anyone teach me?

### Website URL | Problem website

https://waline.felixesintot.top

### Where your waline deploy? | Where is the service deployed?

Vercel (Default)

### Where your comment data store? | Where is the data stored?

LeanCloud(https://leancloud.app)

**Comment by lizheming** [comment, marked as answer]
1. 昵称和邮箱不为空可直接配置见文档 https://waline.js.org/reference/client/props.html#requiredmeta
2. 注册用户昵称不可重复 demo，具体可参照这个代码逻辑自己调试改下

```js
// index.js
const Waline = require('@waline/vercel');

module.exports = Waline({
  plugins: [
    {
      middlewares: [
        async (ctx, next) => {
          if(ctx.method !== 'POST' || ctx.url !== '/api/user') {
            return next();
          }
      
         const { storage } = think.config();
          const userModel = ctx.services(`storage/${storage}`, 'Users');
          const { display_name } = ctx.post();

          const users = await userModel.where({ display_name }).select();
          if (think.isEmpty(users)) {
              ctx.status = 500;
              return ctx.fail('昵称被占用');
          }

          await next();
        },
      ],
    },
  ],
});
```

**Reply by felixesintot** [reply]
太感谢了

---
Thank you so much

**Reply by felixesintot** [reply]
可是这个程序不起作用

---
But this program doesn’t work

**Reply by lizheming** [reply]
> 具体可参照这个代码逻辑自己调试改下

---
> For details, you can refer to this code logic to debug and modify it yourself

**Reply by felixesintot** [reply]
提示 ctx.services 不是一个函数？

---
Tip: ctx.services is not a function?

**Reply by felixesintot** [reply]
实在搞不动了……

能不能 DEBUG 也帮我做一下，真的不会

---
I really can’t do it anymore…

Can DEBUG also help me do it? I really don’t know how

**Reply by lizheming** [reply]
多写了个s，改成 `ctx.service` 

---
I wrote an extra s and changed it to `ctx.service`

**Reply by felixesintot** [reply]
这又是啥：Cannot read properties of undefined (reading '_complex')

**Reply by felixesintot** [reply]
```javascript
              const userModel = ctx.service(`storage/${storage}`,'Users');
              const users = await userModel.where({display_name: display_name,mail: ['!=', email]}).select();
```
这两行死活调试不出来

---
```javascript
              const userModel = ctx.service(`storage/${storage}`,'Users');
              const users = await userModel.where({display_name: display_name,mail: ['!=', email]}).select();
```
These two lines cannot be debugged

**Reply by felixesintot** [reply]
算了，最后我自己在程序中链接 leancloud 也解决了（~~果然 DIY 还是用自己熟悉的东西比较好~~）

---
Forget it, I finally solved it by linking leancloud in the program (~~It’s better to DIY with something you are familiar with~~)

**Reply by felixesintot** [reply]
此贴结，仍然感谢 Waline 的开发者们

---
This post ends, thanks to the developers of Waline
